### ISSUE TITLE
test: test_substr_with_null_values fails in Clickhouse

### ISSUE THREAD

**Issue by datapythonista** [ibis-project/ibis#2642]
`test_substr_with_null_values` is failing in Clickhouse. The test was not running in Azure by mistake, and the failure has been detected and xfailed while moving the job to GitHub actions. See #2624

```
___________________ test_substr_with_null_values[TestConf0] ____________________

backend = <ibis.backends.clickhouse.tests.conftest.TestConf object at 0x7f4a3ddb3190>
alltypes = ref_0
ClickhouseDatabaseTable[table]
  name: ibis_testing.`functional_alltypes`
  schema:
    date : date[non-nullable...2[non-nullable]*] 'year' from table
      ref_0
    month = Column[int32[non-nullable]*] 'month' from table
      ref_0
df =            date  index  Unnamed: 0  ...       timestamp_col  year  month
0    2009-01-01   3898        3898  ... 2009-...5:08:13  2010     12
7299 2010-12-31   3277        3277  ... 2010-12-31 05:09:13  2010     12

[7300 rows x 16 columns]

    @pytest.mark.xfail_unsupported
    def test_substr_with_null_values(backend, alltypes, df):
        table = alltypes.mutate(
            substr_col_null=ibis.case()
            .when(alltypes['bool_col'], alltypes['string_col'])
>           .else_(None)
            .end()
            .substr(0, 2)
        )

ibis/backends/tests/test_string.py:283: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ibis/expr/operations.py:1390: in end
    return op.to_expr()
ibis/expr/operations.py:119: in to_expr
    self._expr_cached = self._make_expr()
ibis/expr/operations.py:123: in _make_expr
    klass = self.output_type()
ibis/expr/operations.py:1481: in output_type
    exprs = self.results + [self.default]
ibis/expr/types.py:1031: in __add__
    return type(self.op())(self.values + other_values).to_expr()
ibis/expr/operations.py:119: in to_expr
    self._expr_cached = self._make_expr()
ibis/expr/operations.py:123: in _make_expr
    klass = self.output_type()
ibis/expr/operations.py:3073: in output_type
    dtype = rlz.highest_precedence_dtype(self.values)
ibis/expr/rules.py:39: in highest_precedence_dtype
    return dt.highest_precedence(expr.type() for expr in exprs)
ibis/expr/datatypes.py:1621: in highest_precedence
    return functools.reduce(higher_precedence, dtypes)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

left = String(nullable=False), right = Null(nullable=True)

    def higher_precedence(left: DataType, right: DataType) -> DataType:
        if castable(left, right, upcast=True):
            return right
        elif castable(right, left, upcast=True):
            return left
    
        raise com.IbisTypeError(
>           'Cannot compute precedence for {} and {} types'.format(left, right)
        )
E       ibis.common.exceptions.IbisTypeError: Cannot compute precedence for string[non-nullable] and null types

ibis/expr/datatypes.py:1615: IbisTypeError
```

**Comment by cpcloud** [comment]
Duplicate of #3135

**Comment by gforsyth** [comment]
Duplicate of #2891 (#3135 was closed in favor of #2891)
